module/data/item/templates/activated-effect.mjs

```diff
@@ -107,6 +107,21 @@ export default class ActivatedEffectTemplate extends SystemDataModel {
     return { value: this.uses.value, max: this.parent.system.uses.max, name: "system.uses.value" };
   }
 
+  /* -------------------------------------------- */
+
+  /**
+   * Prepare activated effect data.
+   */
+  prepareActivatedEffectData() {
+    if ( !this.parent.isEmbedded ) return;
+    if ( ["ammo", "charges", "material"].includes(this.consume.type) && this.consume.target.includes(".") ) {
+      const item = this.parent.actor.items.find(i =>
+        (i.flags.core?.sourceId === this.consume.target) || (i.flags.dnd5e?.sourceId === this.consume.target)
+      );
```

Comment:
I think we need to build up a `sourceId` map when doing the initial `prepareEmbeddedItems` pass, and then perform these lookups in `prepareFinalAttributes` probably. I don't think we can afford to scan over the entire Item list for multiple Items (though I'm sure it's entirely possible that we are inadvertently doing this elsewhere).

Alternatively, we could perform the lookup only once when the Item is added to the Actor for the first time (in `Item5e#_onCreate` maybe) and set the reference at that point. It has the disadvantage that Items can't be added out-of-order, (i.e. you must add Ki before you add Flurry of Blows) but I would lean fairly strongly towards this solution if it doesn't pose any immediate issues (like with Advancement addition/removal/modification).